ICU-11276 Adding NumberRangeFormatter for number, currency, and unit range formats

sffc opened the pull request:
##### Checklist

- [ ] Issue filed at https://unicode-org.atlassian.net :  ICU-______
- [ ] Update PR title to include Issue number
- [ ] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

aheninger commented on icu4c/source/i18n/unicode/numberrangeformatter.h:
```diff
@@ -0,0 +1,839 @@
+     * Overload of numberFormatterSecond() for use on an rvalue reference.
+     *
+     * @param formatterSecond
+     *            The formatter to use for both numbers in the range.
```
both -> the second

aheninger commented on icu4c/source/i18n/unicode/numberrangeformatter.h:
```diff
@@ -0,0 +1,839 @@
+     * @draft ICU 63
+     * @see Appendable
+     */
+    Appendable &appendTo(Appendable &appendable, UErrorCode& status);
```
Why is this function not const?

sffc replied:
Discussed.  Also see ICU-20110

aheninger commented on icu4c/source/i18n/unicode/numberrangeformatter.h:
```diff
@@ -0,0 +1,839 @@
+    Appendable &appendTo(Appendable &appendable, UErrorCode& status);
+
+    /**
+     * Determines the start and end indices of the next occurrence of the given <em>field</em> in the
```
Should probably be explicit on whether the end index is inclusive or exclusive.

sffc commented:
@aheninger This is ready for review.  Java and C++ are both implemented.  The tests may or may not pass due to the lack of some data, but they pass locally with patched data.

macchiati commented on icu4j/main/classes/core/src/com/ibm/icu/number/FormattedNumberRange.java:
```diff
@@ -0,0 +1,209 @@
+     * @see Appendable
+     * @see NumberRangeFormatter
+     */
+    public <A extends Appendable> A appendTo(A appendable) {
```
I suggest elevating this into a general utility, since we could use it all over the place, not just with numbers.

sffc replied:
Hmm.  There is no base class for this.  I suppose we could make a pure-virtual class (interface in Java) called `FormattedValue` with the four main functions:

- toString()
- appendTo()
- nextFieldPosition()
- toCharacterIterator()

Then both this class (FormattedNumberRange) and FormattedNumber would want to implement that interface.

It saves on APIDocs and cleans things up a bit.  It would also be an API change, though, and I'm not sure if I see a strong use case for this particular class being abstracted.  Since this is the second implementation, this would be the time to add such a parent class.  I'll send an email; it's worth asking.

sffc replied:
I realized that there is more design work required for this.  I filed a ticket:

https://unicode-org.atlassian.net/browse/ICU-20138

I think we should leave this as-is for now because adding a new parent class shouldn't break source compatibility.

aheninger commented on icu4c/source/i18n/unicode/numberrangeformatter.h:
```diff
@@ -0,0 +1,851 @@
+     */
+    UnlocalizedNumberRangeFormatter() = default;
+
+    // Make default copy constructor call the NumberFormatterSettings copy constructor.
```
This non-doc comment is out-of-place here.

sffc replied:
Moved the comment to the implementation files, here and elsewhere.

sffc commented on icu4c/source/i18n/numrange_fluent.cpp:
```diff
@@ -0,0 +1,428 @@
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_FORMATTING && !UPRV_INCOMPLETE_CPP11_SUPPORT
```
Done.

sffc commented on icu4c/source/i18n/numrange_impl.cpp:
```diff
@@ -0,0 +1,348 @@
+    // this part of the code should be updated to load from the local numbering system.
+    // The numbering system could come from the one specified in the NumberFormatter passed to
+    // numberFormatterBoth() or similar.
+
```
Filed a ticket and put the ticket ID here in the code.

sffc commented on icu4c/source/i18n/numrange_impl.h:
```diff
@@ -0,0 +1,89 @@
+
+
+/**
+ * Implementation class for UFormattedNumber with magic number for safety.
```
Fixed.

sffc commented:
@aheninger Changes since your last review:

1. Rebased the PR to master, resolving conflicts
2. Added two new commits, test cases and minor changes:
    - 432f7ec
    - a7c78e1

I will merge this in as soon as I have another approval.
